tfs/frame.py

```diff
@@ -233,33 +236,48 @@ def validate(
     data_frame: TfsDataFrame | pd.DataFrame,
     info_str: str = "",
     non_unique_behavior: str = "warn",
+    compatibility: str = "madx",
 ) -> None:
     """
-    Check if a data frame contains finite values only, strings as column names and no empty headers
-    or column names.
+    Enforce validity rules on a `TfsDataFrame` (see admonition below).
+    Additional checks are performed for compatibility with either ``MAD-X``
+    or ``MAD-NG`` as provided by the `compatibility` parameter.
+
+    .. admonition:: Methodology
+
+        This function performs several different checks on the provided dataframe. The following
+        checks are performed for all compatibility modes (``MAD-X`` and ``MAD-NG``):
 
-    .. admonition:: **Methodology**
+          1. Checking no single element in the data is a `list` or `tuple`.
+          2. Checking for non-physical values in the dataframe (uses `.isna()` with the right option context).
+          3. Checking for duplicates in either indices or columns.
+          4. Checking for column names that are not strings.
+          5. Checking for column names including spaces.
 
-        This function performs several different checks on the provided dataframe:
-         1. Checking no single element is a `list` or `tuple`, which is done with a
-            custom vectorized function applied column-by-column on the dataframe.
-         2. Checking for non-physical values in the dataframe, which is done by
-            applying the ``isna`` function with the right option context.
-         3. Checking for duplicates in either indices or columns.
-         4. Checking for column names that are not strings.
-         5. Checking for column names including spaces.
+        When checking for ``MAD-X`` compatibility, which is more restrictive than ``MAD-NG``,
+        the following additional checks are performed:
 
+          1. Checking no boolean values are in the dataframe headers.
+          2. Checking no complex values are in the dataframe headers.
+          3. Checking no boolean-dtype columns are in the dataframe.
+          4. Checking no complex-dtype columns are in the dataframe.
 
     Args:
         data_frame (TfsDataFrame | pd.DataFrame): the dataframe to check on.
         info_str (str): additional information to include in logging statements.
+        compatibility (str): Which code to check for compatibility with. Accepted values
+            are `madx`, `mad-x`, `madng` and `mad-ng`, case-insensitive. Defauts to `madx`.
         non_unique_behavior (str): behavior to adopt if non-unique indices or columns are found in the
             dataframe. Accepts `warn` and `raise` as values, case-insensitively, which dictates
             to respectively issue a warning or raise an error if non-unique elements are found.
     """
+    if compatibility.lower() not in VALIDATION_MODES:
+        errmsg = f"Invalid compatibility mode provided: '{compatibility}'."
+        raise ValueError(errmsg)
```

Comment:
**Note:** in the `validate` function directly we expect `compatibility` to be one of the valid values. `None` is not accepted here, only in reader / writer functions (which will then skip calling `validate`).